File: src/main/java/dev/wnuke/NukeClient/module/ModuleManager.java
package dev.wnuke.NukeClient.module;

import dev.wnuke.NukeClient.util.ModuleSettings;
import org.reflections.Reflections;

import java.lang.reflect.InvocationTargetException;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import static dev.wnuke.NukeClient.NukeClient.SETTINGS_MANAGER;


public class ModuleManager {
    private Reflections reflections = new Reflections("dev.wnuke");
    public static Map<String, Module> modules = new HashMap<>();
    public void loadModules() throws NoSuchMethodException, IllegalAccessException, InvocationTargetException, InstantiationException {
        modules.clear();

        // Find and register all modules
        Set<Class<? extends Module>> moduleClasses = reflections.getSubTypesOf(Module.class);
        for (Class<? extends Module> moduleClass : moduleClasses) {
            Module module = moduleClass.getConstructor().newInstance();
            modules.put(module.getName(), module);
        }

        // Load or create config
        SETTINGS_MANAGER.loadSettings();

        // Enable modules if their settings indicate they should be
        for (Map.Entry<String, Module> module : modules.entrySet()) {
            if ((boolean) module.getValue().getSettings().getSetting("enabled")) {
                module.getValue().onEnable();
            }
        }
    }
    public ModuleSettings getGlobalSettings() {
        return modules.get("Global Settings").getSettings();
    }

    public Map<String, Module> getModules() {
        return modules;
    }
}

File: src/main/kotlin/dev/wnuke/nukeclient/module/ModuleManager.kt
package dev.wnuke.nukeclient.module

import org.reflections.Reflections
import kotlin.reflect.KClass

class ModuleManager {
    private val modules: HashSet<Module> = HashSet()

    init {
        modules.clear()
        val moduleClassIterator =
                Reflections("dev.wnuke.nukeclient.module").getSubTypesOf(
                        Module::class.java
                ).iterator()
        while (moduleClassIterator.hasNext()) {
            val module = moduleClassIterator.next().getConstructor().newInstance()
            modules.add(module)
        }
    }

    fun <T:Module> getModuleByClass(clazz: KClass<T>): Module? {
        val moduleIterator = modules.iterator()
        while (moduleIterator.hasNext()) {
            val current = moduleIterator.next()
            if (current::class == clazz) return current
        }
        return null
    }

    fun getModuleByName(name: String): Module? {
        val moduleIterator = modules.iterator()
        while (moduleIterator.hasNext()) {
            val current = moduleIterator.next()
            if (current.name == name) return current
        }
        return null
    }

    fun getModulesInCat(category: Category): HashSet<Module> {
        val modulesInCategory: HashSet<Module> = HashSet()
        val moduleIterator = modules.iterator()
        while (moduleIterator.hasNext()) {
            val current = moduleIterator.next()
            if (current.category == category) modulesInCategory.add(current)
        }
        return modulesInCategory
    }
}
